src/libexpr/paths.cc

```diff
@@ -12,4 +13,10 @@ SourcePath EvalState::rootPath(PathView path)
     return {rootFS, CanonPath(absPath(path))};
 }
 
+SourcePath EvalState::stringWithContextToPath(std::string_view s, const NixStringContext & context)
+{
+    auto path = CanonPath(s);
+    return !context.empty() ? SourcePath{storeFS, std::move(path)} : rootPath(std::move(path));
```

Comment:
This should probably validate two rules:

- If it has a context, it needs to start with storeDir. (otherwise, we're probably looking at a script or data with a reference)
- The string needs to be an absolute path. It looks like `rootPath(CanonPath)` just eats garbage if not absolute, and the other overload would be worse, using the process' working directory.

Reply:
Unfortunately, this runs into a problem: `EvalState::mkOutputStringRaw()` produces paths that don't include `/nix/store` for some reason (e.g. `/125qgp92...`). (It renders a `DownstreamPlaceholder` and to be honest I don't know what that is...)

@Ericson2314 Apparently in 3a7b330b64c6ea77e18a0a96aad7fb14947382d9 you removed the store prefix from these paths. Do you remember why that is? It seems wrong, since generally we only want to rewrite hashes and not change the lengths of strings.

Reply:
- https://github.com/NixOS/nix/issues/12361